Documentation/designs/status-subresource.md

```diff
@@ -0,0 +1,436 @@
+* `ShardStatuses` which is a drilled-down version for each Prometheus shard.
+* [TBD] Number of resources (service monitors, pod monitors, probes and Prometheus rules) selected/rejected by the operator.
+
+```golang
+type PrometheusStatus struct {
+	// Represents whether any actions on the underlying managed objects are
+	// being performed. Only delete actions will be performed.
+	Paused bool `json:"paused"`
+	// Total number of non-terminated pods targeted by this Prometheus deployment
+	// (their labels match the selector).
+	Replicas int32 `json:"replicas"`
+	// Total number of non-terminated pods targeted by this Prometheus deployment
+	// that have the desired version spec.
+	UpdatedReplicas int32 `json:"updatedReplicas"`
+	// Total number of available pods (ready for at least minReadySeconds)
+	// targeted by this Prometheus deployment.
+	AvailableReplicas int32 `json:"availableReplicas"`
+	// Total number of unavailable pods targeted by this Prometheus deployment.
+	UnavailableReplicas int32 `json:"unavailableReplicas"`
+	// The current state of the Prometheus service.
+	// +patchMergeKey=type
+	// +patchMergeStrategy=merge
+	// +optional
+	Conditions []PrometheusStatusCondition `json:"conditions,omitempty"`
+	// The list has one entry per shard. Each entry provides a summary of the shard status.
+	// +patchMergeKey=shardID
+	// +patchMergeStrategy=merge
+	// +optional
+	ShardStatuses []ShardStatus `json:"shardStatuses,omitempty"`
+}
+
+// PrometheusStatusCondition represents the state of the resources associated with the Prometheus resource.
+// +k8s:deepcopy-gen=true
+type PrometheusStatusCondition struct {
+	// Type of the condition being reported.
+	// +required
+	Type PrometheusConditionType `json:"type"`
+	// status of the condition.
+	// +required
+	Status PrometheusConditionStatus `json:"status"`
+	// lastTransitionTime is the time of the last update to the current status property.
+	// +required
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+	// Reason for the condition's last transition.
+	// +optional
+	Reason *string `json:"reason,omitempty"`
+	// Human-readable message indicating details for the condition's last transition.
+	// +optional
+	Message *string `json:"reason,omitempty"`
+}
+
+type PrometheusConditionType string
+
+const (
+	// Available indicates that enough Prometheus pods are ready to provide the service.
+	PrometheusAvailable PrometheusConditionType = "Available"
+	// Degraded indicates that some Prometheus pods don't run as expected.
+	PrometheusDegraded PrometheusConditionType = "Degraded"
+	// Reconciled indicates that the operator has reconciled the state of the underlying resources with the Prometheus object spec.
+	PrometheusReconciled PrometheusConditionType = "Reconciled"
```

Comment:
My expectation from a condition named this way would be that it goes to false as soon as the spec is updated, and the controller sets it to true at the end of the reconciliation loop. However, I can also see the use of the condition that you're describing, so maybe it's just a matter of coming up with better naming. 

Alternatively, a status field instead of a condition could be more suitable for this purpose. From what I understand, conditions are supposed to represent calculated/inferred values so that other controllers don't need to recalculate values from a resource's status.

From the k8s [api conventions](https://github.com/kubernetes/community/blob/4c9ef2d/contributors/devel/sig-architecture/api-conventions.md): 
```
Conditions provide a standard mechanism for higher-level status reporting from a controller. 
They are an extension mechanism which allows tools and other controllers to collect 
summary information about resources without needing to 
understand resource-specific status details
```

Reply:
> My expectation from a condition named this way would be that it goes to false as soon as the spec is updated, and the controller sets it to true at the end of the reconciliation loop. However, I can also see the use of the condition that you're describing, so maybe it's just a matter of coming up with better naming.

Right, I see how the current name can be confusing/misleading. I still think that it makes sense to convey to the user if the operator has applied the current spec and if not, what failed.

Reply:
Yes, absolutely makes sense to have this information in the status.